roadmap/roadmap_2024.md

```diff
@@ -0,0 +1,105 @@
+faces significant deprecations that impact automatic certificate generation.
+For more insights into these challenges, see [Issue #3524](https://github.com/kubernetes-sigs/kubebuilder/issues/3524).
+
+This situation necessitates a reevaluation of its inclusion and potentially prompts users to
+adopt alternatives like cert-manager by default. Additionally, the requirement to manually rebuild
+[kube-rbac-proxy images—due](https://github.com/kubernetes-sigs/kubebuilder/blob/master/RELEASE.md#to-build-the-kube-rbac-proxy-images)
+to its external status from Kubernetes-SIG—places a considerable maintenance
+burden on Kubebuilder maintainers.
+
+**Motivations:**
+- Address kube-rbac-proxy breaking changes/deprecations.
+  - For further information: [Issue #3524 - kube-rbac-proxy warn about deprecation and future breaking changes](https://github.com/kubernetes-sigs/kubebuilder/issues/3524)
+- Feedback from the community has highlighted a preference for cert-manager's default integration, aiming security with Prometheus and metrics.
+  - More info: [GitHub Issue #3524 - Improve scaffolding of ServiceMonitor](https://github.com/kubernetes-sigs/kubebuilder/issues/3657)
+- Desire for kube-rbac-proxy to be optional, citing its prescriptive nature.
+  - See: [Issue #3482 - The kube-rbac-proxy is too opinionated to be opt-out.](https://github.com/kubernetes-sigs/kubebuilder/issues/3482)
+- Reduce the maintainability effort to generate the images used by Kubebuilder projects and dependency within third-party solutions.
+  - Related issues:
+    - [Issue #1885 - use a NetworkPolicy instead of kube-rbac-proxy](https://github.com/kubernetes-sigs/kubebuilder/issues/1885)
+    - [Issue #3230 - Migrate away from google.com gcp project kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/issues/3230)
+
+**Proposed Solutions:**
+
+- **Making kube-rbac-proxy Optional:** Offering users the option to include kube-rbac-proxy caters to diverse project
+requirements and simplifies the transition towards its potential externalization or removal,
+reducing future maintenance efforts.
+
+- **Leveraging NetworkPolicies:** This alternative focuses on minimizing external dependencies by
+utilizing Kubernetes-native solutions like NetworkPolicies, in line with our maintenance reduction goals.
+
+- **Default Enablement of cert-manager:** While not directly addressing the maintenance concerns related to
+kube-rbac-proxy, defaulting to cert-manager responds to community feedback and navigates the upcoming deprecations.
+This strategy also acknowledges cert-manager's existing role as a prerequisite for webhooks.
+
+**We encourage the Kubebuilder community to participate in this discussion, offering feedback and contributing ideas
+to refine these proposals. Your involvement is crucial in shaping the future of secure and efficient project scaffolding in Kubebuilder.**
+
```

Comment:
Some topics have come to mind from the previous discussions in the community:

### 1\. Server Side Apply (SSA)

This issue has a [long history](https://kubernetes.slack.com/archives/CAR30FCJZ/p1645534614215639) and appears to be [blocked](https://github.com/kubernetes-sigs/kubebuilder/issues/2514) due to challenges with upstream components. However, the controller-runtime team seems to be actively working on `SSA` recently. See:

1.  [kubernetes/kubernetes#118138](https://github.com/kubernetes/kubernetes/issues/118138)
2.  [controller-runtime#2702](https://github.com/kubernetes-sigs/controller-runtime/pull/2702)

Perhaps we'll have the opportunity to address this within the year?

### 2\. Updating Controller Scaffolding to Align with the Latest Interface from controller-runtime

Since `v3` of Kubebuilder (kb), we've provided a stable scaffolding layout to facilitate controller development.
However, some interfaces from `controller-runtime` appear outdated, and recent versions have introduced new structures/interfaces.
I'm wondering if it's time to reevaluate our scaffolding code template and consider trimming the use of legacy structures/interfaces if necessary?

Reply:
Good points. I added the item 2 but not 1.
The `Server Side Apply (SSA)` for now still an objetive of Controller-Tools
We only will be able to have it as objetive for us when they implement that

Also, I think we need to try to keep the list small
Not more than 5 goals in the year.
We have no available effort to looking for all. But I either agree that the most import goals needs to be there.

Reply:
Glad to hear that the information on `SSA` and the annual goals was helpful. Thanks for the clarification!

Reply:
Closing this one but feel free to re-open
REALLY THANK YOU FOR ALL YOUR SUPPORT AND COLLAB !!